### PR TITLE
Post Status Control

### LINKED ISSUE #40: Add post status dropdown selection control and preview
When opening a draft post in the Customizer, it would be ideal to be able to make changes and then _publish_ the post from there. For example in a workflow to preview how changes to a title wraps: https://youtu.be/sXu2pA42J88

As part of this, we can also implement the Trash button (#137) alongside the post status dropdown.

### LINKED ISSUE #137: Delete a post whilst in the Customizer
It would be great if I could trash or delete a post whilst in the Customizer.

I recognize that this task isn't small, and will need some UX exploration, etc. before it can go to development.

### PR DESCRIPTION
**Pull request by valendesigns**
Fixes #40, fixes #137

### CODE REVIEW

File: js/customize-preview-posts.js [hunk 1]
```diff
@@ -150,6 +150,10 @@
 	};
 
 	api.bind( 'preview-ready', function() {
+		api.bind( 'saved', function() {
+			api.preview.send( 'customized-posts-purge-trash', api.previewPosts.data.isSingular );
+		} );
```
Comment (westonruter) [inline js/customize-preview-posts.js]: @valendesigns I think this isn't right. The `saved` event is triggered on the `api` inside the controls itself. Why send this `customized-posts-purge-trash` message from the preview?

Why not just do in `customize-posts.js`:

```js
api.bind( 'saved', function() {
    component.purgeTrash();
} );
```

File: js/customize-post-section.js [hunk 2]
```diff
@@ -179,6 +179,9 @@
 			if ( postTypeObj.supports.title || postTypeObj.supports.slug ) {
 				section.addSlugControl();
 			}
+
+			section.addPostStatusControl();
```
Comment (westonruter) [inline js/customize-post-section.js]: @valendesigns since sync isn't supported yet, maybe we should just skip adding the `post_status` control altogether if `EditPostPreviewCustomize` is defined?

File: js/customize-post-section.js [hunk 3]
```diff
@@ -281,6 +284,96 @@
...
+						'default': setting.id
+					},
+					field_type: 'select',
+					setting_property: 'post_status',
+					choices: api.Posts.data.postStatusChoices
+				}
+			} );
+
+			/**
+			 * Update the UI when a post is trasitioned from/to trash.
```
Comment (westonruter) [inline js/customize-post-section.js]: Typo `s/trasitioned/transitioned/`

File: js/customize-preview-posts.js (hunk 1 above)
Comment (valendesigns) [inline js/customize-preview-posts.js]: `is_singular()` always returned `false` when called from the pane. So I needed access to that value to force the preview to reload the archive.

File: js/customize-post-section.js [hunk 4]
```diff
@@ -281,6 +284,96 @@
...
+			 *
+			 * @param {bool} trashed - Whether or not the post_status is 'trash'.
```
Comment (westonruter) [inline js/customize-post-section.js]: `s/bool/boolean/` PhpStorm likes `boolean` over `bool`. It looks like `bool` is just for PHP. It's not mentioned in JSDoc: http://usejsdoc.org/tags-type.html

File: js/customize-post-section.js (hunk 2 above)
Comment (valendesigns) [inline js/customize-post-section.js]: Sounds good. How do we access that from `customize-post-section.js`

File: js/customize-preview-posts.js (hunk 1 above)
Comment (westonruter) [inline js/customize-preview-posts.js]: I think it would be better to keep track of whatever the last data sent from `customized-posts` was instead.

```js
component.previewedQuery = new api.Value( {} );
api.previewer.bind( 'customized-posts', function ( data ) {
    var query = {};
    _.each( [ 'isSingular', 'isPostPreview', 'queriedPostId' ] ).each( function( key ) {
        query[ key ] = data[ key ];
    } );
    component.previewedQuery.set(  ) );    
} );
```

Then at any time you can access whether the preview `isSingular` via `query.previewedQuery.get().isSingular`.

This would be a basic implementation of https://core.trac.wordpress.org/ticket/36582

File: js/customize-post-section.js (hunk 2 above)
Comment (westonruter) [inline js/customize-post-section.js]: Just check if `typeof EditPostPreviewCustomize !== 'undefined'`.

File: php/class-wp-customize-post-setting.php [hunk 5]
```diff
@@ -508,12 +503,23 @@ protected function update( $data ) {
...
-			return false;
+		if ( $is_trashed ) {
+			add_filter( 'wp_insert_post_empty_content', '__return_false', 100 );
 		}
-		return true;
+		$r = wp_update_post( wp_slash( $data ), true );
+		$result = is_wp_error( $r );
+
+		if ( $result && $is_trashed ) {
+			$result = wp_trash_post( $is_trashed );
```
Comment (westonruter) [inline php/class-wp-customize-post-setting.php]: This needs to be:

```php
$result = wp_trash_post( $this->post_id );
```

**Comment by westonruter** [comment]
- [ ] Add test to ensure a post with an empty title, content, and excerpt can be trashed.
- [ ] Add test to ensure inserted auto-draft post has the expected guid.
- [ ] Add test for `filter_customize_save_response_to_export_saved_values`.

File: php/class-wp-customize-post-setting.php [hunk 6]
```diff
@@ -307,15 +306,15 @@ public function sanitize( $post_data, $strict = false ) {
 		$post_data = sanitize_post( $post_data, 'db' );
 		$initial_sanitized_post_data = $post_data;
 
-		$maybe_empty = 'attachment' !== $this->post_type
+		$maybe_empty = 'trash' !== $post_data['post_status'] && 'attachment' !== $this->post_type
```
Comment (westonruter) [inline php/class-wp-customize-post-setting.php]: I think this is wrong. We should instead be doing this change below:

```php
if ( 'trash' !== $post_data['post_status'] && apply_filters( 'wp_insert_post_empty_content', $maybe_empty, $post_data ) ) {
```
